popup.js

```diff
@@ -0,0 +1,107 @@
+}
+
+function getCsrfToken(origin) {
+    return fetch(csrfFetchUrlForOrigin(origin), { credentials: 'same-origin' })
+    .then(res => res.json())
+    .then(res => res.query.tokens.csrftoken);
+}
+
+function getDefaultListId(url) {
+    return fetch(getReadingListsUrlForOrigin(url.origin), { credentials: 'same-origin' })
+    .then(res => {
+        if (!res.ok) {
+            return setUpReadingListsForUser(url)
+            .then(res => getDefaultListId(url));
+        }
+        return res.json()
+        .then(res => res.lists.filter(list => list.default)[0].id);
+    })
+}
+
+function setUpReadingListsForUser(url) {
+    return getCsrfToken(url.origin)
+    .then(token => fetch(readingListSetupUrlForOrigin(url.origin, token), { method: 'POST', credentials: 'same-origin' }));
+}
+
+function parseTitleFromUrl(url) {
+    return url.searchParams.has('title') ? url.searchParams.get('title') : url.pathname.replace('/wiki/', '');
+}
+
+function show(id) {
+    // Use setTimeout to work around an extension popup resizing bug on Chrome
+    // see https://bugs.chromium.org/p/chromium/issues/detail?id=428044
+    setTimeout(() => { document.getElementById(id).style.display = 'block' }, 50);
+}
+
+function showLoginPage(url) {
+    browser.tabs.update({ url: `${url.origin}/wiki/Special:UserLogin?returnto=${parseTitleFromUrl(url)}` });
+}
+
+function showLoginPrompt(url) {
+    document.getElementById('loginButton').onclick = (element) => showLoginPage(url);
+    show('loginPromptContainer');
+}
+
+function showAddToListSuccessMessage() {
+    show('addToListSuccessContainer');
+}
+
+function showAddToListFailureMessage(res) {
+    document.getElementById('failureReason').textContent = res.detail ? res.detail : res.title ? res.title : res.type ? res.type : res;
+    show('addToListFailedContainer');
+}
+
+function mobileToCanonicalHost(url) {
+    url.hostname = url.hostname.replace(/^m\./, '').replace('.m.', '.');
+    return url;
+}
+
+function getAddToListPostBody(url) {
+    return `project=${mobileToCanonicalHost(url).origin}&title=${parseTitleFromUrl(url)}`;
+}
+
+function getAddToListPostOptions(url) {
+    return {
+        method: 'POST',
+        headers: { 'content-type': 'application/x-www-form-urlencoded' },
+        credentials: 'same-origin',
+        body: getAddToListPostBody(url)
+    }
+}
+
+function handleAddPageToListResult(res) {
+    if (res.id) showAddToListSuccessMessage(); else showAddToListFailureMessage(res);
+}
+
+function addPageToDefaultList(url, listId, token) {
+    return fetch(readingListPostEntryUrlForOrigin(url.origin, listId, token), getAddToListPostOptions(url))
+    .then(res => res.json())
+    .then(res => handleAddPageToListResult(res));
+}
+
+function handleTokenResult(url, token) {
+    return token === '+\\' ? showLoginPrompt(url) : getDefaultListId(url).then(listId => addPageToDefaultList(url, listId, token));
```

Comment:
I wrote it this way for development just to bail out as early as possible, but yeah, this approach will have to go away when we incorporate more translated messages from the API.   I've got a local WIP branch that does this that I'll push up to my fork and link here in a bit.

Note: On https://phabricator.wikimedia.org/T191817 I've proposed creating a message more specific to reading lists for this scenario since the notloggedin message currently returned ("You must be logged in to edit your private information") seemed overly generic.

Reply:
Okay, here's roughly how I envision that looking: https://github.com/mdholloway/readinglists-extension-chrome/commit/f44c08b31f62418c6cf0d1bdc3f601f330f13ae6

In summary I add handling for the `notloggedin` case when hitting the API for the default list ID and display the message from that response if found.  (Actually, I was wrong about the error detail returned in that case, it's "You must be logged in to *view* your private information," which is probably worse for an "add to reading list" extension.)